spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/SGD.scala

```diff
@@ -16,52 +16,78 @@
-class SGD[@specialized(Float, Double) T: ClassTag](implicit ev: TensorNumeric[T])
+class SGD[@specialized(Float, Double) T: ClassTag](
+  var learningRate: Double = 1e-3,
+  var learningRateDecay: Double = 0.0,
+  var weightDecay: Double = 0.0,
+  var momentum: Double = 0.0,
+  var dampening: Double = Double.MaxValue,
+  var nesterov: Boolean = false,
+  var learningRates: Tensor[T] = null,
+  var weightDecays: Tensor[T] = null)(implicit ev: TensorNumeric[T])
   extends OptimMethod[T] {
 
   import SGD._
 
+  if (this.dampening == Double.MaxValue) this.dampening = this.momentum
+
+  var clr = learningRate
+
+  override val config = T(
+    "learningRate" -> this.learningRate,
+    "learningRateDecay" -> this.learningRateDecay,
+    "weightDecay" -> this.weightDecay,
+    "momentum" -> this.momentum,
+    "dampening" -> this.dampening,
+    "nesterov" -> this.nesterov,
+    "learningRates" -> this.learningRates,
+    "weightDecays" -> this.weightDecays,
+    "clr" -> this.clr
+  )
+
   /**
    *
    * @param feval a function that takes a single input (X), the point of a evaluation,
    * and returns f(X) and df/dX
    * @param x the initial point
-   * @param config a table with configuration parameters for the optimizer
-   * learningRate: learning rate
-   * learningRateDecay: learning rate decay
-   * weightDecay: weight decay
-   * weightDecays: 1D tensor of individual weight decays
-   * momentum: momentum
-   * dampening: dampening for momentum
-   * nesterov: enables Nesterov momentum
-   * learningRates: 1D tensor of individual learning rates
    * @param state a table describing the state of the optimizer; after each call the state
    * is modified
    * @return the new x 1D tensor and the function list, evaluated before the update
    */
   override def optimize(feval: (Tensor[T]) => (T, Tensor[T]), x: Tensor[T],
-    config: Table, state: Table = null): (Tensor[T], Array[T]) = {
+                        state: Table): (Tensor[T], Array[T]) = {
 
-    val _state = if (state == null) config else state
-    val lrSchedule = config.get[LearningRateSchedule]("learningRateSchedule").getOrElse(Default())
-    lrSchedule.updateHyperParameter(config, _state)
+    val _state = state
+    val lrSchedule = _state.get[LearningRateSchedule]("learningRateSchedule").getOrElse(Default())
+    lrSchedule.updateHyperParameter(this, _state)
 
-    val wd = config.get[Double]("weightDecay").getOrElse(0.0)
-    val mom = config.get[Double]("momentum").getOrElse(0.0)
-    val damp = config.get[Double]("dampening").getOrElse(mom)
-    val nesterov = config.get[Boolean]("nesterov").getOrElse(false)
-    val lrs = config.get[Tensor[T]]("learningRates").getOrElse(null)
-    val wds = config.get[Tensor[T]]("weightDecays").getOrElse(null)
+    val wd = this.weightDecay
+    val mom = this.momentum
+    val damp = this.dampening
+    val nesterov = this.nesterov
+    val lrs = this.learningRates
+    val wds = this.weightDecays
+    val clr = ev.fromType(this.clr)
```

Comment:
yeah, no clr in SGD now. clr is produced by leaningrate schedule. So clr is defined in leaningrate schedule.